source/Ldap.Integration.Tests/scripts/ActiveDirectory/Azure/main.tf

```diff
@@ -0,0 +1,170 @@
+  resource_group_name = azurerm_resource_group.rg.name
+}
+
+## <https://www.terraform.io/docs/providers/azurerm/r/subnet.html> 
+resource "azurerm_subnet" "subnet" {
+  name                 = "subnet"
+  resource_group_name  = azurerm_resource_group.rg.name
+  virtual_network_name = azurerm_virtual_network.vnet.name
+  address_prefixes       = ["10.0.2.0/24"]
+}
+
+resource "azurerm_public_ip" "public_ip" {
+  name                = "public_ip"
+  resource_group_name = azurerm_resource_group.rg.name
+  location            = azurerm_resource_group.rg.location
+  allocation_method   = "Dynamic"
+}
+
+## <https://www.terraform.io/docs/providers/azurerm/r/network_interface.html>
+resource "azurerm_network_interface" "example" {
+  name                = "nic1"
+  location            = azurerm_resource_group.rg.location
+  resource_group_name = azurerm_resource_group.rg.name
+
+
+  ip_configuration {
+    name                          = "internal"
+    subnet_id                     = azurerm_subnet.subnet.id
+    private_ip_address_allocation = "Dynamic"
+    public_ip_address_id = azurerm_public_ip.public_ip.id 
+  }
+}
+
+# Create Network Security Group and rule
+resource "azurerm_network_security_group" "myterraformnsg" {
+    name                = "securitygroup"
+    location            = azurerm_resource_group.rg.location
+    resource_group_name = azurerm_resource_group.rg.name
+
+    security_rule {
+        name                       = "LDAP"
+        priority                   = 1002
+        direction                  = "Inbound"
+        access                     = "Allow"
+        protocol                   = "Tcp"
+        source_port_range          = "*"
+        destination_port_range     = "389"
+        source_address_prefix      = "*"
+        destination_address_prefix = "*"
+    }
+}
+
+resource "azurerm_network_interface_security_group_association" "example" {
+    network_interface_id      = azurerm_network_interface.example.id
+    network_security_group_id = azurerm_network_security_group.myterraformnsg.id
+}
+
+
+resource "random_password" "password" {
+  length           = 16
+  special          = true
+  override_special = "_%@"
+}
+
+## <https://www.terraform.io/docs/providers/azurerm/r/windows_virtual_machine.html>
+resource "azurerm_windows_virtual_machine" "example" {
+  name                = "ad-vm"
+  resource_group_name = azurerm_resource_group.rg.name
+  location            = azurerm_resource_group.rg.location
+  size                = "Standard_F2"
+  admin_username      = "adminuser"
+  admin_password      = random_password.password.result
+  network_interface_ids = [
+    azurerm_network_interface.example.id,
+  ]
+
+  os_disk {
+    caching              = "ReadWrite"
+    storage_account_type = "Standard_LRS"
+  }
+
+  source_image_reference {
+    publisher = "MicrosoftWindowsServer"
+    offer     = "WindowsServer"
+    sku       = "2016-Datacenter"
+    version   = "latest"
+  }
```

Comment:
Consider baking the AD setup into a packer script and reference that image output here. VM extensions can be pretty brutal for setting this sort of thing up; despite your use of `depends_on` I don't know if Az will guarantee that they run in order or necessarily succeed before returning control back to Terraform. If you aren't able to use a golden image, consider using Ansible or similar to set up the VM once terraform's provisioned it, that should be more reliable.

Reply:
This _does_ tend to work, the only complexity is that the AD Service install requires a reboot which then causes the configuration scripts not to run through the `CustomScriptExtension` resource.
Hence the `RunCommandWindows` that comes afterwards.

Reply:
It may well have worked for you a small number of times, but depending on how frequent this is run, the tolerance for failure may be quite low. Generally for this sort of thing I'd recommend baking as much setup as possible into a VM image.